pennylane/templates/subroutines/iqp.py

```diff
@@ -0,0 +1,135 @@
+            `[[[0]], [[1]], [[2]], [[3]]]` specifies a circuit with single qubit rotations on the first four qubits, each
+            with its own trainable parameter. The `pattern` `[[[0],[1]], [[2],[3]]]` correspond to a circuit with two
+            trainable parameters with generators :math:`X_0+X_1` and :math:`X_2+X_3` respectively. A circuit with a
+            single trainable gate with generator :math:`X_0\otimes X_1` corresponds to the `pattern`
+            `[[[0,1]]]`.
+        spin_sym (bool, optional): If True, the circuit is equivalent to one where the initial state
+            :math:`\frac{1}{\sqrt(2)}(|00\dots0> + |11\dots1>)` is used in place of :math:`|00\dots0>`.
+
+    Raises:
+        Exception: when pattern and weights have a different number of elements.
+    """
+
+    resource_keys = {"spin_sym", "pattern", "num_wires"}
+
+    def __init__(
+        self, weights, num_wires, pattern, spin_sym=None, id=None
+    ):  # pylint: disable=too-many-arguments
+        if len(pattern) != len(weights):
+            raise ValueError(
+                f"Number of gates and number of parameters for an Instantaneous Quantum Polynomial circuit must be the same, got {len(pattern)} gates and {len(weights)} weights."
+            )
+
+        if num_wires == 0:
+            raise ValueError("At least one valid wire is required.")
+
+        self._hyperparameters = {
+            "spin_sym": spin_sym,
+            "weights": weights,
+            "pattern": pattern,
+        }
+        super().__init__(wires=range(num_wires), id=id)
+
+    @classmethod
+    def _primitive_bind_call(cls, *args, **kwargs):
+        return cls._primitive.bind(*args, **kwargs)
+
+    @property
+    def resource_params(self):
+        return {
+            "spin_sym": self.hyperparameters["spin_sym"],
+            "pattern": self.hyperparameters["pattern"],
+            "num_wires": len(self.wires),
+        }
+
+
+def _instantaneous_quantum_polynomial_resources(spin_sym, pattern, num_wires):
+    resources = defaultdict(int)
+    if spin_sym:
+        resources[
+            resource_rep(
+                PauliRot,
+                pauli_word="Y" + "X" * (num_wires - 1),
+            )
+        ] = 1
+
+    resources[resource_rep(Hadamard)] = 2 * num_wires
+
+    for gate in pattern:
+        for gen in gate:
+            resources[resource_rep(MultiRZ, num_wires=len(gen))] += 1
+
+    return resources
+
+
+@register_resources(_instantaneous_quantum_polynomial_resources)
+def _instantaneous_quantum_polynomial_decomposition(
+    wires, weights, pattern, spin_sym
+):  # pylint: disable=unused-argument, too-many-arguments
+    num_wires = len(wires)
+
+    if spin_sym:
+        PauliRot(2 * np.pi / 4, "Y" + "X" * (num_wires - 1), wires=range(num_wires))
+
+    for i in range(num_wires):
+        Hadamard(i)
+
+    for par, gate in zip(weights, pattern):
+        for gen in gate:
+            MultiRZ(2 * par, wires=gen)
```

Comment:
Can you remind me if the `MultiRZ` gates all commute with each other such that the order doesn't matter?